percona/tls/tls.go

```diff
@@ -0,0 +1,100 @@
+package tls
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"fmt"
+	"math/big"
+	"time"
+)
+
+var validityNotAfter = time.Date(9999, 12, 31, 23, 59, 59, 0, time.UTC)
+
+// Issue returns CA certificate, TLS certificate and TLS private key
+func Issue(hosts []string) (caCert []byte, tlsCert []byte, tlsKey []byte, err error) {
+	rsaBits := 2048
+	priv, err := rsa.GenerateKey(rand.Reader, rsaBits)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("generate rsa key: %w", err)
+	}
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("generate serial number for root: %w", err)
+	}
+	subject := pkix.Name{
+		Organization: []string{"Root CA"},
+	}
+	caTemplate := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               subject,
+		NotBefore:             time.Now(),
+		NotAfter:              validityNotAfter,
+		KeyUsage:              x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+	}
+
+	derBytes, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &priv.PublicKey, priv)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("generate CA certificate: %w", err)
+	}
+	certOut := &bytes.Buffer{}
+	err = pem.Encode(certOut, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("encode CA certificate: %w", err)
+	}
+	cert := certOut.Bytes()
+
+	serialNumber, err = rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("generate serial number for client: %w", err)
+	}
+	subject = pkix.Name{
+		Organization: []string{"PGO"},
+	}
+	issuer := pkix.Name{
+		Organization: []string{"Root CA"},
+	}
+	tlsTemplate := x509.Certificate{
+		SerialNumber:          serialNumber,
+		Subject:               subject,
+		Issuer:                issuer,
+		NotBefore:             time.Now(),
+		NotAfter:              validityNotAfter,
+		DNSNames:              hosts,
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth},
+		BasicConstraintsValid: true,
+		IsCA:                  false,
```

Comment:
We're setting `isCA: true` if we create the key pair using cert-manager in https://github.com/percona/percona-postgresql-operator/blob/97f05063d6d3acc395d332ae6a1fd9fde67a9b1a/percona/controllers/pgc/tls.go#L162

Reply:
https://github.com/percona/percona-postgresql-operator/pull/199/commits/2493db43abbe08183d7a8c34001a3579b5899ef7